test/jdk/javax/swing/JFileChooser/FileSystemView/SystemIconTest.java

```diff
@@ -72,9 +72,20 @@ static void negativeTests() {
 
     static void testSystemIcon(File file, boolean implComplete) {
         int[] sizes = new int[] {16, 32, 48, 64, 128};
+        if (!file.exists() || !file.canRead()) {
```

Comment:
If this file is not accessible the "getSystemIcon" should return "null" you can check it here.

Reply:
> If this file is not accessible the "getSystemIcon" should return "null" you can check it here.

If file is not accessible we still return default icon for file or folder - if we, for example, looking at the folder in file explorer  with file not readable then we should display some icon for it. It will not be icon from the file itself - because we can not read it - but it will not be null.

Reply:
1. If the file does not exist then the getSystemIcon() will always return null, but what happens if the file cannot be read? We will call the "sf = ShellFolder.getShellFolder" and then we call "sf.getIcon()" and it returns what? As far as I understand it always return MultiResolutionIconImage, no?

2. If we return some icon when the file cannot be read then you can check it here as well.

Reply:
> * but what happens if the file cannot be read? We will call the "sf = ShellFolder.getShellFolder" and then we call "sf.getIcon()" and it returns what? As far as I understand it always return MultiResolutionIconImage, no?
> *

No, the native code will fail to load the icon but since file does exist and we only return null for nonexistent file we will fall back to UIManager.getIcon which returns icon from the installed LAF resource bundle. This is correct behavior since otherwise we might end in file manager with file that has no icon at all. Right now that will be a single resolution icon for all the LAFs. Again, in the future that can be changed but right now it is a single resolution icon.

Reply:
> No, the native code will fail to load the icon
So what call will fail? Is the "sf.getIcon()" return null? in what line of code?

Reply:
Yes, from my remote debugging it happens in sf.getIcon() but the circumstances are quite rare - file has to be inaccessible and the user's GUI session should not be initialized - that's why this failure only happens on headless testing. Another solution would be to mark this test as headful - that would solve issue as well.

Reply:
I am a bit surprised the tests for this are not already headful. I can imagine all sorts of platform reasons why making them headless is not testing the "real" use case making the tests less than useful.
Ever seen a hidpi  headless system ??

Reply:
> I am a bit surprised the tests for this are not already headful

Fixed.

Reply:
That test was not marked as headful because this API should work in the headless environment as well, if it does not work when we have a bug in the shell classes which should return the icon from the UIManager on any error. I think it worked as expected.

As far as I understand when we cannot access the image of some specific file we tried to get a generic image via shell32.dll, does it fail as well? Is it possible that when we try to create the failback icon here:
https://github.com/openjdk/jdk/blob/82bfc5d45c54fb37dc021bc91fa17efe34f77f44/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1199
we get the icon of the correct size and skip the creation of "MultiResolutionIconImage"? As a result, the test fails because the icon is not multi-resolution.

Reply:
> we get the icon of the correct size and skip the creation of "MultiResolutionIconImage"? As a result, the test fails because the icon is not multi-resolution.

Well, it might be this case sometimes but in the case when i was able to reproduce this issue it was null coming from native code so we were falling back to UIManager. I can enforce MRI creation even if size is correct - but again, the method signature does not promise the MRI in all cases so we might just modify the test case to not expect the MRI as granted even for the well-known files.

Reply:
When in what line of code we got a null from the native code so the UIManager was used?

Reply:
> When in what line of code we got a null from the native code so the UIManager was used?

I haven't got into the exact native part which is failing. The problem is that i was able to reproduce it only twice in a couple of weeks - both time remotely, both time on headless machine. If i insert enough debug output to pinpoint the location it stops reproducing.

Reply:
> That test was not marked as headful because this API should work in the headless environment as well

Ok, i removed the headful mark and also removed the testing for icon always be MRI on windows because it is not so under some circumstances and it is within spec to return single resolution icon so this assumption is incorrect.

Reply:
> so under some circumstances and it is within spec to return single resolution icon 

Under what circumstances? Above we found a code path that may incorrectly return the non-MRI images. Do we have some other similar places?

Reply:
> Do we have some other similar places?

No, i do not see any other places where we return direct result of OS giving out an icon.

Reply:
Then we can just fix this one place, and check how it will work.